app/controllers/repositories_controller.rb

```diff
@@ -104,14 +108,50 @@ def add_like # MAKE A LIKE CONTROLLER TO PUT THIS IN
       render json: { failed: true }
   end
 
+  def password_entry
+  end
+
+  def pass_authenticate
+    @auth = Repository.authenticate(params[:slug], params[:password])
+    @repository = Repository.find_by_slug(params[:slug])
+    respond_to do |format|
+      if @auth
+        @authorized = true
+        authorized_repo_ids << params[:id]
+        flash[:notice] = "Success"
+        format.html{redirect_to repository_path}
+        format.json { render json: { role: :guest }, status: :ok }
+      else
+        @authorized = false
+        flash[:notice] = "Incorrect password. Try again!"
+        format.html { redirect_to password_entry_repository_path}
+        format.json { render json: @repository.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
   private
 
+    def check_session
+      if authorized_repo_ids.include? params[:id]
+        @authorized = true
+      end
+    end
+
+    def check_auth
+      if (@authorized == true || (@user.admin?) || (@user.staff?) || (params[:user_username] == @user.username))
```

Comment:
Can this line be bypassed by a hacker? I can control `params[:user_username]` so that it matches my username. I think you need to check that `@user.username == @repository.user_username`

Reply:
done